kvpress/presses/duo_attention_press.py

```diff
@@ -0,0 +1,104 @@
+# SPDX-FileCopyrightText: Copyright (c) 1993-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+from io import StringIO
+from dataclasses import dataclass
+from contextlib import contextmanager
+
+import torch
+import requests  # type: ignore[import-untyped]
+import numpy as np
+
+from kvpress.presses.base_press import BasePress
+
+
+PATTERNS_DICT = {
+    "togethercomputer/Llama-2-7B-32K-Instruct": "Llama-2-7B-32K-Instruct/lr%3D0.02-reg%3D0.05-ctx%3D1000_32000-multi_passkey10",  # noqa: E501
+    "gradientai//Llama-3-8B-Instruct-Gradient-1048k": "Llama-3-8B-Instruct-Gradient-1048k/lr%3D0.02-reg%3D0.05-ctx%3D1000_32000-multi_passkey10",  # noqa: E501
+    "gradientai//Llama-3-8B-Instruct-Gradient-4194k": "Llama-3-8B-Instruct-Gradient-4194k/lr%3D0.02-reg%3D0.05-ctx%3D1000_32000-multi_passkey10",  # noqa: E501
+    "meta-llama/Meta-Llama-3.1-8B-Instruct": "Meta-Llama-3.1-8B-Instruct/lr=0.02-reg=0.05-ctx=1000_128000-multi_passkey10",  # noqa: E501
+    "mistralai/Mistral-7B-Instruct-v0.2": "Mistral-7B-Instruct-v0.2/lr%3D0.02-reg%3D0.05-ctx%3D1000_32000-multi_passkey10",  # noqa: E501
+    "mistralai/Mistral-7B-Instruct-v0.3": "Mistral-7B-Instruct-v0.3/lr%3D0.02-reg%3D0.05-ctx%3D1000_32000-multi_passkey10",  # noqa: E501
+}
+
+
+@dataclass
+class DuoAttentionPress(BasePress):
+    """
+    Implements DuoAttention (https://arxiv.org/abs/2410.10819)
+
+    Splits attention heads into two types:
+    - Retrieval heads: use the full KV cache
+    - Streaming heads: use only sink and recent tokens.
+
+    Head classification is based on scores loaded from https://github.com/mit-han-lab/duo-attention/
+    The higher the head_compression_ratio, the more streaming heads are used.
+    """
+
+    head_compression_ratio: float = 0.0
```

Comment:
Could you define attributes used by the class in `__post_init__`  (e.g. `self.compression_ratio_`) ?